File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
class RandomizedSet {
    unordered_map<int,int> mp;
    vector<int> v;
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if(mp.find(val)!= mp.end()) return false;
        v.push_back(val);
        mp[val]= v.size()-1;
        return true;
    }
    
    bool remove(int val) {
        if(mp.find(val)==mp.end()) return false;
        int end= v.back();
        mp[end]= mp[val];
        v[mp[val]] = end;
        v.pop_back();
        mp.erase(val);
        return true;
    }
    
    int getRandom() {
        return v[rand()%v.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
class RandomizedSet {

    HashMap<Integer, Integer> mp = new HashMap<>();
    ArrayList<Integer> nums = new ArrayList<>();
    public RandomizedSet() {
        mp = new HashMap<>();
        nums = new ArrayList<>();
    }
    
    public boolean insert(int val) {
        if(mp.containsKey(val)) return false;
        mp.put(val, nums.size());
        nums.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if(mp.containsKey(val)==false) return false;
        int idx = mp.get(val);
        if(idx < nums.size()-1){
            mp.put(nums.get(nums.size()-1), idx);
            nums.set(idx, nums.get(nums.size()-1));
        }
        mp.remove(val);
        nums.remove(nums.size()-1);
        return true;
    }
    
    public int getRandom() {
        Random rand = new Random();
        return nums.get(rand.nextInt(nums.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */